Library/OcLogAggregatorLib/OcLog.c

```diff
@@ -363,24 +366,51 @@ InternalLogAddEntry (
     // Write to internal buffer.
     //
 
+    OldAsciiBufferOffset = Private->AsciiBufferOffset;
+
     Status = AsciiStrCatS (Private->AsciiBuffer, Private->AsciiBufferSize, Private->TimingTxt);
     if (!EFI_ERROR (Status)) {
-      Status = AsciiStrCatS (Private->AsciiBuffer, Private->AsciiBufferSize, Private->LineBuffer);
+      Private->AsciiBufferOffset += AsciiStrLen (Private->TimingTxt);
+      Status                      = AsciiStrCatS (Private->AsciiBuffer, Private->AsciiBufferSize, Private->LineBuffer);
+      if (!EFI_ERROR (Status)) {
+        Private->AsciiBufferOffset += AsciiStrLen (Private->LineBuffer);
+      }
     }
 
     //
     // Write to a file.
-    // Always overwriting file completely is most reliable.
-    // I know it is slow, but fixed size write is more reliable with broken FAT32 driver.
     //
     if (((OcLog->Options & OC_LOG_FILE) != 0) && (OcLog->FileSystem != NULL)) {
       if (EfiGetCurrentTpl () <= TPL_CALLBACK) {
-        OcSetFileData (
-          OcLog->FileSystem,
-          OcLog->FilePath,
-          Private->AsciiBuffer,
-          (UINT32)Private->AsciiBufferSize
-          );
+        if (OcLog->UnsafeLogFile != NULL) {
+          //
+          // For non-broken FAT32 driver this is fine. For driver with broken write
+          // support (e.g. Aptio IV) this can result in corrupt file or unusable fs.
+          //
+          WriteSize   = Private->AsciiBufferOffset - OldAsciiBufferOffset;
+          WrittenSize = WriteSize;
+          OcLog->UnsafeLogFile->Write (OcLog->UnsafeLogFile, &WrittenSize, &Private->AsciiBuffer[OldAsciiBufferOffset]);
+          OcLog->UnsafeLogFile->Flush (OcLog->UnsafeLogFile);
+          if (WriteSize != WrittenSize) {
+            DEBUG ((
+              DEBUG_VERBOSE,
+              "OCL: Log write truncated %u to %u\n",
+              WriteSize,
+              WrittenSize
+              ));
```

Comment:
Can this recurse?

Reply:
Hmm, I guess not, there are multiple verbose debug statements in OcSetFileData as used in the safe path.